src/spikeinterface/postprocessing/template_similarity.py

```diff
@@ -60,25 +85,102 @@ def _get_data(self):
 
 
-def compute_similarity_with_templates_array(templates_array, other_templates_array, method):
+def compute_similarity_with_templates_array(templates_array, other_templates_array, method, n_shifts, mask=None):
+
     import sklearn.metrics.pairwise
 
-    if method == "cosine_similarity":
+    all_metrics = ["cosine", "l1", "l2"]
+
+    if method in all_metrics:
+        nb_templates = templates_array.shape[0]
         assert templates_array.shape[0] == other_templates_array.shape[0]
-        templates_flat = templates_array.reshape(templates_array.shape[0], -1)
-        other_templates_flat = templates_array.reshape(other_templates_array.shape[0], -1)
-        similarity = sklearn.metrics.pairwise.cosine_similarity(templates_flat, other_templates_flat)
+        n = templates_array.shape[1]
+        nb_templates = templates_array.shape[0]
+        assert n_shifts < n, "max_lag is too large"
+        num_shifts = 2 * n_shifts + 1
+        distances = np.ones((num_shifts, nb_templates, nb_templates), dtype=np.float32)
+        if mask is not None:
+            units_overlaps = np.sum(mask, axis=2) > 0
+            overlapping_templates = {}
+            for i in range(nb_templates):
+                overlapping_templates[i] = np.flatnonzero(units_overlaps[i])
+
+        # We can use the fact that dist[i,j] at lag t is equal to dist[j,i] at time -t
+        # So the matrix can be computed only for negative lags and be transposed
+
+        for count, shift in enumerate(range(-n_shifts, 1)):
+            if mask is None:
+                src_templates = templates_array[:, n_shifts : n - n_shifts].reshape(nb_templates, -1)
+                tgt_templates = templates_array[:, n_shifts + shift : n - n_shifts + shift].reshape(nb_templates, -1)
+                if method == "l1":
+                    norms_1 = np.linalg.norm(src_templates, ord=1, axis=1)
+                    norms_2 = np.linalg.norm(tgt_templates, ord=1, axis=1)
+                    denominator = norms_1[:, None] + norms_2[None, :]
+                    distances[count] = sklearn.metrics.pairwise.pairwise_distances(
+                        src_templates, tgt_templates, metric="l1"
+                    )
+                    distances[count] /= denominator
+                elif method == "l2":
+                    norms_1 = np.linalg.norm(src_templates, ord=2, axis=1)
+                    norms_2 = np.linalg.norm(tgt_templates, ord=2, axis=1)
+                    denominator = norms_1[:, None] + norms_2[None, :]
+                    distances[count] = sklearn.metrics.pairwise.pairwise_distances(
+                        src_templates, tgt_templates, metric="l2"
+                    )
+                    distances[count] /= denominator
+                else:
+                    distances[count] = sklearn.metrics.pairwise.pairwise_distances(
+                        src_templates, tgt_templates, metric=method
+                    )
```

Comment:
I thik I would prefer explicitly method = "cosine" instead of relying of method from sklearn

Reply:
you mean just having: `sklearn.metrics.pairwise.pairwise_distances(..., method="cosine")`?

Reply:
done